Clarify taint effects in the documentation.

bmcustodio opened the pull request:
As part of a previous PR, `NoExecute` started being recommended as the
effect that should be placed on nodes to avoid unmanaged pods. While
this is correct and required for guaranteeing in the best possible way
pods don't come up as unmanaged, there are some considerations that
can be made and trade-offs that can be pointed out. This PR attempts
at providing a clarification on the taint-based approach to prevent
unmanaged pods by adding a page describing the known implications of
each effect.

Refs #18486.

```release-note
Clarify taint effects in the documentation.
```

ldelossa reviewed (approved):
💯

itsthatriver commented on Documentation/gettingstarted/taints.rst:
```diff
@@ -0,0 +1,87 @@
+  unmanaged and have their networking being managed by another CNI plugin.
+
+* If ``NoExecute`` is used, pods won't be *executed* on a node until Cilium is
+  has the change to remove the taint. One practical effect of this is that
```
```suggestion
* If ``NoExecute`` is used, pods won't be *executed* on a node until Cilium
  has the chance to remove the taint. One practical effect of this is that
```

bmcustodio replied:
Nice catch, thank you!

itsthatriver reviewed (commented):
Noticed a couple typos. I like the writeup, I think this does a good job of clarifying the tradeoffs. For. a bit of context, we're rolling out Cilium to GKE, EKS, and AKS, and because of the complexities of AKS in particular (specifically the required system pool), we assume we can't address all issues with taints and run a separate control loop process to check for unmanaged pods and terminate them, causing them to be recreated and generally get managed (strictly speaking eventually consistent, in practice it tends to happen very infrequently when used in conjunction with `NoSchedule` taints placed everywhere we can put them).

christarazi commented on Documentation/gettingstarted/taints.rst:
```diff
@@ -0,0 +1,87 @@
+why ``NoExecute`` is recommended, as assuming the taint is added back in this
+scenario, already-scheduled pods won't run.
+
+However, on some environments or cloud providers, and as mentioned above, it may
+happen that a taint established at the node-pool level is added back to a node
+after Cilium has removed it and for reasons other than a node upgrade/reset.
+The exact circumstances in which this may happen may vary, but this may lead to
+unexpected/undesired pod evictions in the particular case when ``NoExecute`` is
+being used as the taint effect. It is, thus, recommended that in each deployment
+and depending on the environment or cloud provider, a careful decision is made
+regarding the taint effect (or even regarding whether to use the taint-based
+approach at all) based on the information above, on the environment or cloud
+provider's documentation, and on the fact that one is essentially establishing
+a trade-off between having unmanaged pods in the cluster (which can lead to
+dropped traffic and other issues) and having unexpected/undesired evictions
+(which can lead to application downtime).
```
Seems like this paragraph is essentially referencing https://github.com/cilium/cilium/issues/17631. Might be worth mentioning it.

bmcustodio replied:
I'm not 100% sure — it may be in some situations, and it's a bit unpredictable because we don't have a lot of visibility into what's happening over time in a given cloud provider and across the different cloud providers, but the paragraph actually and essentially tries to address the case in which a taint is added back to a node, causing pods to be evicted/deleted. So I am actually not sure if, in all of these cases, the CEPs are kept.

aanm commented on Documentation/gettingstarted/taints.rst:
```diff
@@ -0,0 +1,87 @@
+* Some cloud providers may revert changes made to the CNI configuration by
+  Cilium during operations such as node reboots, updates or routine maintenance.
+
+This is notably the case with GKE and some other cloud providers, in which node
```
Might worth be more specific on which cloud providers this does happen. On AKS with its BYO CNI won't happen. On GKE with dataplane v2 this also doesn't happen for example.

bmcustodio replied:
I think GKE non-Dataplane v2 is a good way to put it. I'm not sure if other cloud providers actually reset parts of the filesystem as GKE does. We never faced these issues in AWS, for example, and I'm not sure if AKS does reset the CNI configuration back on node reboots. So far, the only evidence is GKE (non-Dataplane V2). I'm rewording this slightly to make it more evident. :+1:

aanm commented on Documentation/gettingstarted/taints.rst:
```diff
@@ -0,0 +1,87 @@
+  concurrently with Cilium when the node next reboots, and hence may become
+  unmanaged and have their networking being managed by another CNI plugin.
+
+* If ``NoExecute`` is used, pods won't be *executed* on a node until Cilium has
```
Let's add something like:

_won't be executed on a node ..._ but they will be scheduled [1].

and then [1] can point to https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/

bmcustodio replied:
Hmmm they won't really be scheduled:

> If there is at least one un-ignored taint with effect NoExecute then the pod will be evicted from the node (if it is already running on the node), and will not be scheduled onto the node (if it is not yet running on the node).

What I think can happen here is changing to "pods won't be *scheduled* nor *executed* on a node" (which I did now). WDYT?

aanm replied:
Looks good!

aanm commented on Documentation/gettingstarted/taints.rst:
```diff
@@ -0,0 +1,87 @@
+provider's documentation, and on the fact that one is essentially establishing
+a trade-off between having unmanaged pods in the cluster (which can lead to
+dropped traffic and other issues) and having unexpected/undesired evictions
+(which can lead to application downtime).
```
I would suggest to add one more paragraph with a conclusion stating that Cilium is deployed with `NoExecuted` by default as we believe it the least disruptive mode that users can use to deploy Cilium on cloud providers.

bmcustodio replied:
Sure, added that :+1: Thank you!

bmcustodio committed:
Clarify taint effects in the documentation.

As part of a previous PR, 'NoExecute' started being recommended as the
effect that should be placed on nodes to avoid unmanaged pods. While
this is correct and required for guaranteeing in the best possible way
pods don't come up as unmanaged, there are some considerations that
can be made and trade-offs that can be pointed out. This PR attempts
at providing a clarification on the taint-based approach to prevent
unmanaged pods by adding a page describing the known implications of
each effect.

Signed-off-by: Bruno M. Custódio <brunomcustodio@gmail.com>
